libvast/src/system/connector.cpp

```diff
@@ -0,0 +1,168 @@
+    case caf::sec::no_downstream_stages_defined:
+    case caf::sec::stream_init_failed:
+    case caf::sec::invalid_stream_state:
+    case caf::sec::unhandled_stream_error:
+    case caf::sec::bad_function_call:
+    case caf::sec::feature_disabled:
+    case caf::sec::cannot_open_file:
+    case caf::sec::socket_invalid:
+    case caf::sec::socket_disconnected:
+    case caf::sec::socket_operation_failed:
+    case caf::sec::unavailable_or_would_block:
+    case caf::sec::malformed_basp_message:
+    case caf::sec::serializing_basp_payload_failed:
+    case caf::sec::redundant_connection:
+    case caf::sec::remote_lookup_failed:
+    case caf::sec::no_tracing_context:
+    case caf::sec::all_requests_failed:
+    case caf::sec::field_invariant_check_failed:
+    case caf::sec::field_value_synchronization_failed:
+    case caf::sec::invalid_field_type:
+    case caf::sec::unsafe_type:
+    case caf::sec::save_callback_failed:
+    case caf::sec::load_callback_failed:
+    case caf::sec::conversion_failed:
+    case caf::sec::connection_closed:
+    case caf::sec::type_clash:
+    case caf::sec::unsupported_operation:
+    case caf::sec::no_such_key:
+    case caf::sec::broken_promise:
+    case caf::sec::connection_timeout:
+    case caf::sec::action_reschedule_failed:
+      return false;
+    case caf::sec::incompatible_versions:
+    case caf::sec::incompatible_application_ids:
+      return true;
+  }
+  return true;
+}
+
+bool is_non_recoverable_error(const caf::error& err) {
+  if (err.category() != caf::type_id_v<caf::sec>)
+    return false;
+  const auto err_code = std::underlying_type_t<caf::sec>{err.code()};
+  auto err_enum = caf::sec{caf::sec::none};
+  if (!caf::from_integer(err_code, err_enum)) {
+    VAST_WARN("unable to retrieve error code for a remote node connection "
+              "error:{}",
+              err);
+    return false;
+  }
+  return is_non_recoverable_error_enum(err_enum);
+}
+} // namespace
+
+connector_actor::behavior_type
+connector(connector_actor::stateful_pointer<connector_state> self,
+          std::size_t max_connection_attempts,
+          caf::timespan connection_retry_delay, vast::port port,
+          std::string host, caf::timespan connection_timeout) {
+  self->state.connection_retry_delay = connection_retry_delay;
+  self->state.port = std::move(port);
+  self->state.host = std::move(host);
+  self->state.node_connection_timeout = connection_timeout;
+  return {[self,
+           max_connection_attempts](atom::connect) -> caf::result<node_actor> {
+            if (self->state.promise.pending())
+              return self->state.promise;
+            self->state.remaining_connection_attempts = max_connection_attempts;
+            self->state.promise = self->make_response_promise<node_actor>();
+            VAST_INFO("client connects to VAST node at {}", self->state.host);
+            self->send(self, atom::internal_v, atom::start_v);
+            return self->state.promise;
```

Comment:
Should this just `self->delegate` instead of using a response promise, which makes the whole logic a lot easier to read?